### PR TITLE
TD-106, TD-133, TD-94 adding seasons, buy in, collecting rewards feature

### PR DESCRIPTION
**Pull request by grujawork**
- [x] https://thalesmarket.atlassian.net/browse/TD-133
- [x] https://thalesmarket.atlassian.net/browse/TD-94
- [x] https://thalesmarket.atlassian.net/browse/TD-106
- [x] https://thalesmarket.atlassian.net/browse/TD-132

### CODE REVIEW

File: contracts/ThalesRoyale/ThalesRoyale.sol [hunk 1]
```diff
@@ -232,9 +245,56 @@ contract ThalesRoyale is Owned, Pausable {
...
+
+    function setClaimTime(uint _claimTime) public onlyOwner {
+        claimTime = _claimTime;
+    }
+
+    function _populateWinnersAndThereRewards(address[] memory alivePlayers) internal {
+        require(finished, "Royale must be finished");
+        require(alivePlayers.length > 0, "There is no alive players left in Royale");
+
+        uint rewardPerPlayer = reward / alivePlayers.length;
```
Comment (grujawork) [inline contracts/ThalesRoyale/ThalesRoyale.sol:257]: what if 10000 TH / 3 alive players = 3333.33333333.... 👃

File: contracts/ThalesRoyale/ThalesRoyale.sol [hunk 2]
```diff
@@ -232,9 +245,56 @@ contract ThalesRoyale is Owned, Pausable {
...
+
+    function setClaimTime(uint _claimTime) public onlyOwner {
+        claimTime = _claimTime;
+    }
+
+    function _populateWinnersAndThereRewards(address[] memory alivePlayers) internal {
+        require(finished, "Royale must be finished");
+        require(alivePlayers.length > 0, "There is no alive players left in Royale");
+
+        uint rewardPerPlayer = reward.divideDecimal(alivePlayers.length);
```
Comment (grujawork) [inline contracts/ThalesRoyale/ThalesRoyale.sol:257]: what if 10000 TH / 3 alive players = 3333.33333333.... 👃

is this we need to use 🤔

Comment (grujawork) [reply contracts/ThalesRoyale/ThalesRoyale.sol]: done.

**Comment by dgornjakovic** [comment]
While you are at this, also consider making a ThalesRoyale that has seasons so that we can have one even every e.g. week without having to redeploy.

File: contracts/ThalesRoyale/ThalesRoyale.sol [hunk 3]
```diff
@@ -232,9 +246,54 @@ contract ThalesRoyale is Owned, Pausable {
...
+
+    function _populateWinnersAndThereRewards(address[] memory alivePlayers) internal {
+        require(finished, "Royale must be finished");
+        require(alivePlayers.length > 0, "There is no alive players left in Royale");
+
+        rewardPerPlayer = reward.div(alivePlayers.length);
+
+        for (uint i = 0; i < alivePlayers.length; i++) {
+            rewardCollected[alivePlayers[i]] = false;
+        }
```
Comment (kirilantevski) [inline contracts/ThalesRoyale/ThalesRoyale.sol:262]: obsolete, they are false at the start. 
Previous comment was to delete this for loop. 
However based on Danijel comment, maybe you would need to make accumulation logic. But that can be done similarly as in the EscrowThales, without loops

Comment (dgornjakovic) [reply contracts/ThalesRoyale/ThalesRoyale.sol:262]: yeah also never do for loops in solidity (I know I have one but it was just for testing purposes and its a read method)

**Comment by grujawork** [comment]
> While you are at this, also consider making a ThalesRoyale that has seasons so that we can have one even every e.g. week without having to redeploy.

added in a JIRA: https://thalesmarket.atlassian.net/browse/TD-106
smaller PR's better PR's 😄

File: test/contracts/ThalesRoyale/royaleMaxAlivePlayersTest.js [hunk 4]
```diff
@@ -48,28 +55,43 @@ contract('ThalesRoyale', accounts => {
...
+					var privateKey = '0x' + id;
+	
+					var wallet = new ethers2.Wallet(privateKey);
+					//console.log('wallet  ' + wallet.address);
+					
+					//await ThalesDeployed.approve(wallet.address, toUnit(1), { from: owner });
+					//await ThalesDeployed.transfer(wallet.address, toUnit(1), { from: owner });
+					//await ThalesDeployed.approve(royale.address, toUnit(1), { from: wallet.address });
+					//await ThalesDeployed.transfer(royale.address, toUnit(1), { from: wallet.address });
+					//await royale.signUp(toUnit(1) , { from: wallet.address });
```
Comment (grujawork) [inline test/contracts/ThalesRoyale/royaleMaxAlivePlayersTest.js:86]: this i'm getting unknown account 🤔

**Review by dgornjakovic — APPROVED** [review]

**Pull request merged**